Sources/Subprocess/Platforms/Subprocess+Linux.swift

```diff
@@ -347,77 +379,225 @@ private extension siginfo_t {
+        let error = SubprocessError(
+            code: .init(.failedToMonitorProcess),
+            underlyingError: .init(rawValue: errno)
+        )
+        _processMonitorState.withLock { storage in
+            storage = .failed(error)
+        }
+        return
+    }
+
+    // Register shutdownFileDescriptor with epoll
+    var event = epoll_event(
+        events: EPOLLIN.rawValue,
+        data: epoll_data(fd: shutdownFileDescriptor)
+    )
+    var rc = epoll_ctl(
+        epollFileDescriptor,
+        EPOLL_CTL_ADD,
+        shutdownFileDescriptor,
+        &event
+    )
+    guard rc == 0 else {
+        let error = SubprocessError(
+            code: .init(.failedToMonitorProcess),
+            underlyingError: .init(rawValue: errno)
+        )
+        _processMonitorState.withLock { storage in
+            storage = .failed(error)
+        }
+        return
+    }
+
+    // Create thread data
+    let context = MonitorThreadContext(
+        epollFileDescriptor: epollFileDescriptor,
+        shutdownFileDescriptor: shutdownFileDescriptor
+    )
+    let threadContext = Unmanaged.passRetained(context)
     // Create the thread. It will run immediately; because it runs in an infinite
     // loop, we aren't worried about detaching or joining it.
-    #if os(FreeBSD) || os(OpenBSD)
-    var thread: pthread_t?
-    #else
     var thread = pthread_t()
-    #endif
-    _ = pthread_create(
-        &thread,
-        nil,
-        { _ -> UnsafeMutableRawPointer? in
-            // Run an infinite loop that waits for child processes to terminate and
-            // captures their exit statuses.
-            while true {
-                // Listen for child process exit events. WNOWAIT means we don't perturb the
-                // state of a terminated (zombie) child process, allowing us to fetch the
-                // continuation (if available) before reaping.
+    rc = pthread_create(&thread, nil, { args in
+        func reportError(_ error: SubprocessError) {
+            _processMonitorState.withLock { state in
+                guard case .started(let storage) = state else {
+                    return
+                }
+                for continuation in storage.continuations.values {
+                    continuation.resume(throwing: error)
+                }
+            }
+        }
+
+        let unmanaged = Unmanaged<MonitorThreadContext>.fromOpaque(args!)
+        let context = unmanaged.takeRetainedValue()
+
+        var events: [epoll_event] = Array(
+            repeating: epoll_event(events: 0, data: epoll_data(fd: 0)),
+            count: 256
+        )
+
+        // Enter the monitor loop
+        monitorLoop: while true {
+            let eventCount = epoll_wait(
+                context.epollFileDescriptor,
+                &events,
+                CInt(events.count),
+                -1
+            )
+            if eventCount < 0 {
+                if errno == EINTR || errno == EAGAIN {
```

Comment:
suggestion: it might be worth introducing a helper function to handle EINTR/EAGAIN since it's such a common pattern throughout this codebase; see https://github.com/apple/swift-system/blob/6ee9a58c36ad98f4bd917a64d153dd211512e65d/Sources/System/Util.swift#L27 for example.